lib/cli/src/commands/run/capabilities/net.rs

```diff
@@ -0,0 +1,306 @@
+    capable: DynVirtualNetworking,
+    unsupported: DynVirtualNetworking,
+}
+
+macro_rules! call {
+    ($self: expr, $fn_name: ident, $( $arg: expr ),* ) => {
+
+        let enable_networking = $self.enable.get_or_init(|| $self.ask_user(stringify!($fn_name)))
+            .map_err(|e| {
+                tracing::error!("{e}");
+                NetworkError::UnknownError
+            })?;
+
+        if enable_networking {
+            return $self.capable.$fn_name( $( $arg ),* ).await;
+        } else {
+            return $self.unsupported.$fn_name( $( $arg ),* ).await;
+        }
+    };
+
+    ($self: expr, $fn_name: ident) => {
+
+        let enable_networking = $self.enable.get_or_init(|| $self.ask_user(stringify!($fn_name))).map_err(|e| {
+                tracing::error!("{e}");
+                NetworkError::UnknownError
+            })?;
+
+        if enable_networking {
+            return $self.capable.$fn_name().await;
+        } else {
+            return $self.unsupported.$fn_name().await;
+        }
+    };
+}
+
+#[derive(Debug, Clone)]
+enum UserSelection {
+    Yes,
+    No,
+    Always,
+}
+
+impl Display for UserSelection {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            UserSelection::Yes => write!(f, "y"),
+            UserSelection::No => write!(f, "n"),
+            UserSelection::Always => write!(f, "a"),
+        }
+    }
+}
+
+impl FromStr for UserSelection {
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
+        let s = s.trim();
+        if s.is_empty() {
+            anyhow::bail!("No input!")
+        }
+
+        let c: char = s.chars().next().unwrap().to_ascii_lowercase();
+
+        Ok(match c {
+            'n' => UserSelection::No,
+            'y' => UserSelection::Yes,
+            'a' => UserSelection::Always,
+            _ => anyhow::bail!("{s} could not be resolved as a selection"),
+        })
+    }
+}
+
+impl AskingNetworking {
+    pub(crate) fn new(pkg_cache_path: PathBuf, capable_networking: DynVirtualNetworking) -> Self {
+        let enable_networking = OnceLock::new();
+
+        Self {
+            enable: enable_networking,
+            capable: capable_networking,
+            unsupported: std::sync::Arc::new(UnsupportedVirtualNetworking::default()),
+            pkg_cache_path,
+        }
+    }
+
+    fn ask_user(&self, fn_name: &str) -> Result<bool> {
+        let theme = ColorfulTheme::default();
+
+        println!("The current package is requesting networking access.");
+        println!("Run the package with `--net` flag to bypass the prompt.");
+        match dialoguer::Input::with_theme(&theme)
+            .with_prompt(format!(
+                "Would you like to allow networking for this package? {}{}",
+                "".bold(),
+                "(yes/no/always)".dimmed()
+            ))
+            .with_initial_text("a")
+            .interact()
```

Comment:
I bet this should be:

```suggestion
            .interact_text()
```

Or perhaps we need to do

```rust
            .default("a") // Instead of with_initial_text
            .interact()
```

Reply:
[e0859cc](https://github.com/wasmerio/wasmer/pull/4975/commits/e0859cc71ee018687e2565c7c4b5b5e4b9b0c6d7)